lib/parser/source/tree_rewriter/action.rb

```diff
@@ -46,10 +46,52 @@ def ordered_replacements
         reps
       end
 
+      def nested_actions
+        reps = []
+        reps << [:wrap, @range, @insert_before, @insert_after] unless @insert_before.empty? &&
+                                                                      @insert_after.empty?
+        reps << [:replace, @range, @replacement] if @replacement
+        reps.concat(@children.flat_map(&:nested_actions))
+      end
+
       def insertion?
         !insert_before.empty? || !insert_after.empty? || (replacement && !replacement.empty?)
       end
 
+      ##
+      # A root action has its range set to the whole source range, even
+      # though it typically do not act on that range.
+      # This method returns the action as if it was a child action with
+      # its range contracted.
+      # @return [Action]
+      def contract
+        raise 'Empty actions can not be contracted' if empty?
+        return self if insertion?
+        range = @range.with(
+          begin_pos: children.first.range.begin_pos,
+          end_pos: children.last.range.end_pos,
+        )
+        with(range: range)
+      end
+
+      ##
+      # @return [Action] that has been moved to the given source_buffer and with the given offset
+      # No check is done on validity of resulting range.
+      def move(source_buffer, offset)
```

Comment:
the name feels like it modifies original object (at least that's what move operator in c++ does 😄 , maybe it's just a wrong association from my side). WDYT about renaming it to `moved` (or something better if you have an idea) to indicate that it doesn't affect receiver.

100% up to you, there's a ton of similar methods in Ruby (like `Hash#merge` and many many others) that are non-destructive.

Reply:
Agreed